File: Work/code/driving_commands.py
import math
import mmap
import os
import struct
import time

FILE_PATH = r"C:\Users\javot\Desktop\sofia_code\shared_data.bin"

FILE_SIZE = 24
SAMPLE_PERIOD = 0.01
DURATION = 30.0

# Ensure the destination folder exists
folder = os.path.dirname(FILE_PATH)

if not os.path.isdir(folder):
    raise FileNotFoundError(
        "Folder does not exist: {}".format(folder)
    )

# Create the binary file with exactly 24 bytes
with open(FILE_PATH, "w+b") as file:
    file.truncate(FILE_SIZE)
    file.flush()

    # Map the entire file
    shared_memory = mmap.mmap(
        file.fileno(),
        length=0,
        access=mmap.ACCESS_WRITE
    )

    sequence = 0
    start_time = time.perf_counter()
    next_sample_time = start_time

    try:
        while time.perf_counter() - start_time < DURATION:
            elapsed_time = time.perf_counter() - start_time

            # Example commands
            # Replace these with your actual values
            delta = 0.05 * math.sin(0.5 * elapsed_time)
            theta_command = 0.0

            # Odd sequence means writing is in progress
            sequence += 1

            shared_memory.seek(0)
            shared_memory.write(
                struct.pack("<Q", sequence)
            )

            # Write two doubles
            shared_memory.seek(8)
            shared_memory.write(
                struct.pack(
                    "<dd",
                    float(delta),
                    float(theta_command)
                )
            )

            # Even sequence means writing is complete
            sequence += 1

            shared_memory.seek(0)
            shared_memory.write(
                struct.pack("<Q", sequence)
            )

            shared_memory.flush()

            print(
                "sequence={}, delta={:.6f}, theta={:.6f}".format(
                    sequence,
                    delta,
                    theta_command
                )
            )

            next_sample_time += SAMPLE_PERIOD
            remaining_time = (
                next_sample_time - time.perf_counter()
            )

            if remaining_time > 0:
                time.sleep(remaining_time)
            else:
                next_sample_time = time.perf_counter()

    except KeyboardInterrupt:
        print("Transmission stopped.")

    finally:
        shared_memory.close()

print("Python transmission finished.")

File: Work/code/driving_commands.m
clear;
clc;

%% Simulation settings
time_step = 0.001;   % Sampling period [s]
duration = 30;       % Duration [s]

trajectory = zeros(ceil(duration / time_step), 2);
commands = [0.0, 0.0];

%% Shared-memory file
filePath = "C:\Users\javot\Desktop\sofia_code\shared_data.bin";

sharedMemory = memmapfile( ...
    filePath, ...
    "Writable", false, ...
    "Format", { ...
        "uint64", [1 1], "Sequence"; ...
        "double", [1 2], "Values" ...
    });

%% Run simulation
startTime = tic;
cnt = 0;

while toc(startTime) < duration

    %% Read a complete shared-memory sample
    validSample = false;

    while ~validSample
        sequenceBefore = sharedMemory.Data.Sequence;

        % Odd sequence means Python is currently writing
        if mod(sequenceBefore, 2) ~= 0
            pause(0.0001);
            continue;
        end

        newCommands = sharedMemory.Data.Values;
        sequenceAfter = sharedMemory.Data.Sequence;

        validSample = ...
            sequenceBefore == sequenceAfter && ...
            mod(sequenceAfter, 2) == 0;
    end

    %% Ensure commands is a 1-by-2 numeric vector
    commands = double(newCommands(:).');

    if numel(commands) ~= 2
        error( ...
            "Expected two command values, but received %d.", ...
            numel(commands));
    end

    delta = commands(1);
    thetaCommand = commands(2);

    %% Run one driving-environment update
    [xNew, yNew] = run_driving_venv(delta, thetaCommand);

    %% Store the returned position
    cnt = cnt + 1;

    if cnt > size(trajectory, 1)
        trajectory(end + 1000, 2) = 0;
    end

    trajectory(cnt, :) = [xNew, yNew];

    pause(time_step);
end

%% Remove unused preallocated rows
trajectory = trajectory(1:cnt, :);

%% Export trajectory
outputScenario = exportToDrivingScenario(trajectory);
